File: backend/src/main/scala/shared/objects/access/ResourceVisibilityPolicy.scala
package shared.objects.access

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

/** 资源可见性策略，只包含默认可见性以及显式授予查看权限的主体列表。 */
final case class ResourceVisibilityPolicy(
  baseAccess: BaseAccess,
  viewerGrants: List[AccessSubject]
)

/** 提供资源可见性策略的 JSON 编解码，字段形状需与前端镜像类型保持一致。 */
object ResourceVisibilityPolicy:
  given Encoder[ResourceVisibilityPolicy] = deriveEncoder[ResourceVisibilityPolicy]
  given Decoder[ResourceVisibilityPolicy] = deriveDecoder[ResourceVisibilityPolicy]

File: frontend/src/objects/shared/access/ResourceVisibilityPolicy.ts
import type { AccessSubject } from './AccessSubject'
import type { BaseAccess } from './BaseAccess'

/** 资源可见性策略；只描述基础可见性以及用户/用户组的查看授权列表。 */
export type ResourceVisibilityPolicy = {
  baseAccess: BaseAccess
  viewerGrants: AccessSubject[]
}

/** 创建默认受限可见性策略；用于不支持资源级管理授权的资源。 */
export function createRestrictedVisibilityPolicy(): ResourceVisibilityPolicy {
  return {
    baseAccess: 'restricted',
    viewerGrants: [],
  }
}
